File: methods/julia/splines/splines.jl
"""
Spline Methods for Derivative Estimation

Implements Dierckx smoothing spline with native derivative support.
"""

# Load common utilities
include("../common.jl")

using Dierckx
using GeneralizedSmoothingSplines
using ForwardDiff

# Import noise estimation module
include(joinpath(@__DIR__, "..", "..", "..", "src", "hyperparameter_selection.jl"))
using .HyperparameterSelection

# ============================================================================
# Dierckx Spline Fitting
# ============================================================================

"""
	fit_dierckx_spline(x, y; k=5, s=nothing, noise_level=0.0)

Fit smoothing spline using Dierckx.
"""
function fit_dierckx_spline(x, y; k = 5, s = nothing, noise_level = 0.0)
	# Auto-select smoothing parameter if not provided
	if isnothing(s)
		n = length(x)
		σ = noise_level * std(y)
		s = n * σ^2  # GCV-like heuristic
	end

	# Fit spline
	spl = Spline1D(x, y; k = k, s = s)

	# Return callable
	return z -> evaluate(spl, z), spl  # Return both func and spline object
end


# ============================================================================
# Method Evaluator (Standard API)
# ============================================================================

"""
	evaluate_dierckx(x, y, x_eval, orders; params=Dict())

Evaluate Dierckx-5 method: Quintic spline with native derivative support.
"""
function evaluate_dierckx(
	x::Vector{Float64},
	y::Vector{Float64},
	x_eval::Vector{Float64},
	orders::Vector{Int};
	params = Dict()
)
	t_start = time()
	predictions = Dict{Int, Vector{Float64}}()
	failures = Dict{Int, String}()

	try
		# Estimate noise from data using wavelet MAD (fair comparison with other methods)
		# Wavelet is robust to signal curvature, unlike diff2 which overestimates for smooth signals
		noise_level = HyperparameterSelection.estimate_noise_wavelet(y)
		_, spl = fit_dierckx_spline(x, y; k = 5, noise_level = noise_level)

		# Dierckx has native derivative support
		for order in orders
			try
				if order == 0
					predictions[order] = [evaluate(spl, xi) for xi in x_eval]
				else
					predictions[order] = [Dierckx.derivative(spl, xi, nu = order) for xi in x_eval]
				end
			catch e
				predictions[order] = fill(NaN, length(x_eval))
				failures[order] = string(e)
			end
		end

		timing = time() - t_start
		return MethodResult("Dierckx-5", "Spline", predictions, failures, timing, true)

	catch e
		@warn "Dierckx-5 failed" exception=e
		timing = time() - t_start
		return MethodResult("Dierckx-5", "Error", Dict(), Dict(0 => string(e)), timing, false)
	end
end


# ============================================================================
# GeneralizedSmoothingSplines (GSS) - High-order smoothing splines
# ============================================================================

"""
	fit_gss_spline(x, y; max_order=7, noise_level=0.0)

Fit smoothing spline using GeneralizedSmoothingSplines.
Uses degree 2p-1 where p is chosen to support derivatives up to max_order.
"""
function fit_gss_spline(x, y; max_order = 7, noise_level = 0.0)
	# Choose p to support max_order derivatives
	# For order n, need degree >= n+1, so 2p-1 >= n+1, thus p >= (n+2)/2
	p = ceil(Int, (max_order + 2) / 2)

	# Create model
	model = GeneralizedSmoothingSplines.SmoothingSpline(p = p, lambda = 0.01)

	# Tune lambda using marginal likelihood if noise level is available
	if noise_level > 0.0
		try
			GeneralizedSmoothingSplines.tune!(model, x, y; show_trace = false)
		catch e
			# If tuning fails, use default lambda
			@warn "GSS tuning failed, using default lambda" exception = e
		end
	end

	# Fit the model
	fitresult, _, _ = GeneralizedSmoothingSplines.MMI.fit(model, 0, x, y)

	# Return prediction function
	pred_fn = function(t::T) where T
		result_vec = GeneralizedSmoothingSplines.MMI.predict(model, fitresult, T[t])
		return result_vec[1]
	end

	return pred_fn, model, fitresult
end

"""
	nth_derivative_forwarddiff(f, x, n)

Compute n-th derivative using recursive ForwardDiff.
Required because TaylorDiff doesn't work with GSS's complex predict implementation.
"""
function nth_derivative_forwarddiff(f, x, n)
	if n == 0
		return f(x)
	elseif n == 1
		return ForwardDiff.derivative(f, x)
	else
		return ForwardDiff.derivative(t -> nth_derivative_forwarddiff(f, t, n - 1), x)
	end
end

"""
	evaluate_gss(x, y, x_eval, orders; params=Dict())

Evaluate GSS method: High-order smoothing spline with automatic lambda tuning.
Supports arbitrary derivative orders using recursive ForwardDiff.
"""
function evaluate_gss(
	x::Vector{Float64},
	y::Vector{Float64},
	x_eval::Vector{Float64},
	orders::Vector{Int};
	params = Dict()
)
	t_start = time()
	predictions = Dict{Int, Vector{Float64}}()
	failures = Dict{Int, String}()

	try
		# Estimate noise from data using wavelet MAD
		noise_level = HyperparameterSelection.estimate_noise_wavelet(y)

		# Fit spline with appropriate degree for max order
		max_order = maximum(orders)
		pred_fn, model, fitresult = fit_gss_spline(x, y; max_order = max_order, noise_level = noise_level)

		# Compute derivatives using recursive ForwardDiff
		for order in orders
			try
				predictions[order] = [nth_derivative_forwarddiff(pred_fn, xi, order) for xi in x_eval]
			catch e
				predictions[order] = fill(NaN, length(x_eval))
				failures[order] = string(e)
			end
		end

		timing = time() - t_start
		return MethodResult("GSS", "Spline", predictions, failures, timing, true)

	catch e
		@warn "GSS failed" exception = e
		timing = time() - t_start
		return MethodResult("GSS", "Error", Dict(), Dict(0 => string(e)), timing, false)
	end
end

File: methods/python/splines/splines.py
"""
Spline-Based Methods for Derivative Estimation

This module implements spline-based derivative estimation methods including:
1. Chebyshev polynomial approximation (fixed and adaptive degree selection)
2. RKHS smoothing splines (m=2 Sobolev penalty)
3. Butterworth lowpass filtering + quintic spline derivatives
4. Finite difference baseline with smoothing + spline
5. SVR regression + quintic spline derivatives
"""

from typing import Dict
import numpy as np
import os

from numpy.polynomial.chebyshev import Chebyshev
from scipy.interpolate import UnivariateSpline
from scipy.signal import butter, filtfilt
from sklearn.svm import SVR

# Import base class from common utilities
import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).parent.parent))
from common import MethodEvaluator, ADAPTIVE_HYPERPARAMS

if ADAPTIVE_HYPERPARAMS:
    from hyperparameters import select_chebyshev_degree


class SplineMethods(MethodEvaluator):
    """
    Spline-based methods for derivative estimation.

    Inherits from base MethodEvaluator and implements spline-specific methods.
    """

    def evaluate_method(self, method_name: str) -> Dict:
        """
        Dispatch to the appropriate spline method based on method name.

        Args:
            method_name: Name of method to evaluate. Supported:
                - "Chebyshev-Basic-Python": Global Chebyshev polynomial (fixed degree)
                - "Chebyshev-AICc": Chebyshev with adaptive degree selection
                - "RKHS_Spline_m2_Python": RKHS smoothing spline (m=2)
                - "Butterworth_Python": Butterworth lowpass + quintic spline
                - "ButterworthSpline_Python": Finite diff + smoothing + spline
                - "SVR_Python": SVR regression + quintic spline

        Returns:
            Dictionary with predictions, failures, and metadata
        """
        if method_name == "Chebyshev-Basic-Python":
            return self._chebyshev()
        elif method_name == "Chebyshev-AICc":
            return self._chebyshev_aicc()
        elif method_name == "RKHS_Spline_m2_Python":
            return self._rkhs_spline_m2()
        elif method_name == "Butterworth_Python":
            return self._butterworth_spline()
        elif method_name == "ButterworthSpline_Python":
            return self._finite_diff_spline()
        elif method_name == "SVR_Python":
            return self._svr_spline()
        else:
            return {
                "predictions": {order: [np.nan] * len(self.x_eval) for order in self.orders},
                "failures": {"error": f"Unknown spline method: {method_name}"}
            }

    def _chebyshev(self) -> Dict:
        """Global Chebyshev polynomial with analytic derivatives.

        Uses fixed degree selection based on training set size.
        For adaptive selection, use Chebyshev-AICc method instead.

        Returns:
            Dictionary containing:
                predictions: Dict mapping derivative order to list of predictions
                failures: Dict of any errors encountered
                meta: Polynomial degree and selection method
        """
        tmin = float(self.x_train.min())
        tmax = float(self.x_train.max())
        n_train = len(self.x_train)

        # Use fixed degree (original behavior)
        # For adaptive selection, use Chebyshev-Adaptive method instead
        deg = max(3, min(20, n_train - 1))
        aicc = None
        selection_method = "fixed"

        poly = Chebyshev.fit(self.x_train, self.y_train, deg=deg, domain=[tmin, tmax])

        predictions = {}
        failures = {}
        for order in self.orders:
            try:
                if order == 0:
                    vals = poly(self.x_eval)
                else:
                    dpoly = poly.deriv(m=order)
                    vals = dpoly(self.x_eval)
                predictions[order] = [float(v) for v in np.asarray(vals)]
            except Exception as e:
                failures[order] = str(e)
                predictions[order] = [np.nan] * len(self.x_eval)

        meta = {"degree": deg, "selection": selection_method}
        if aicc is not None:
            meta["aicc"] = float(aicc)

        return {"predictions": predictions, "failures": failures, "meta": meta}

    def _chebyshev_aicc(self) -> Dict:
        """
        Chebyshev polynomial with adaptive degree selection via AICc.

        Uses Akaike Information Criterion (corrected for small samples) to
        automatically select polynomial degree based on bias-variance tradeoff.

        Returns:
            Dictionary containing:
                predictions: Dict mapping derivative order to list of predictions
                failures: Dict of any errors encountered
                meta: Polynomial degree, AICc score, and selection method
        """
        if not ADAPTIVE_HYPERPARAMS:
            raise RuntimeError("hyperparameters module not available")

        tmin = float(self.x_train.min())
        tmax = float(self.x_train.max())
        n_train = len(self.x_train)

        # Always use adaptive degree selection via AICc
        deg, aicc = select_chebyshev_degree(
            self.x_train, self.y_train,
            max_degree=min(30, n_train - 1),
            min_degree=3
        )

        poly = Chebyshev.fit(self.x_train, self.y_train, deg=deg, domain=[tmin, tmax])

        predictions = {}
        failures = {}
        for order in self.orders:
            try:
                if order == 0:
                    vals = poly(self.x_eval)
                else:
                    dpoly = poly.deriv(m=order)
                    vals = dpoly(self.x_eval)
                predictions[order] = [float(v) for v in np.asarray(vals)]
            except Exception as e:
                failures[order] = str(e)
                predictions[order] = [np.nan] * len(self.x_eval)

        meta = {"degree": deg, "selection": "AICc", "aicc": float(aicc)}
        return {"predictions": predictions, "failures": failures, "meta": meta}

    def _quintic_spline_derivatives(self, x_grid: np.ndarray, y_grid: np.ndarray) -> Dict:
        """Helper: fit quintic spline to (x_grid, y_grid) and return derivatives at self.x_eval.

        This is an internal helper method used by other spline methods to compute derivatives
        from a smoothed or filtered signal.

        Args:
            x_grid: Input grid points
            y_grid: Function values at grid points

        Returns:
            Dictionary containing predictions and failures for all derivative orders
        """
        predictions = {}
        failures = {}
        try:
            us = UnivariateSpline(x_grid, y_grid, k=5, s=0.0)
        except Exception as e:
            # Fall back to cubic if quintic fails
            try:
                us = UnivariateSpline(x_grid, y_grid, k=3, s=0.0)
            except Exception as ee:
                for order in self.orders:
                    failures[order] = f"spline_fit_failed: {ee}"
                    predictions[order] = [np.nan] * len(self.x_eval)
                return {"predictions": predictions, "failures": failures}

        for order in self.orders:
            try:
                d = us.derivative(n=order)
                vals = d(self.x_eval)
                predictions[order] = [float(v) for v in np.asarray(vals)]
            except Exception as e:
                failures[order] = str(e)
                predictions[order] = [np.nan] * len(self.x_eval)
        return {"predictions": predictions, "failures": failures}

    def _rkhs_spline_m2(self) -> Dict:
        """Smoothing spline targeting Sobolev m=2 (penalize integral of f''^2).

        Implemented via UnivariateSpline with smoothing s selected from env or heuristic.
        Higher derivatives obtained via spline's derivative method.

        The smoothing parameter can be controlled via environment variables:
        - RKHS_M2_S: Explicit smoothing parameter value
        - RKHS_NOISE_FLOOR: Noise floor for heuristic (default 0.01)

        Returns:
            Dictionary containing:
                predictions: Dict mapping derivative order to list of predictions
                failures: Dict of any errors encountered
                meta: Smoothing parameter value
        """
        t = self.x_train
        y = self.y_train
        n = len(t)
        s_env = os.environ.get("RKHS_M2_S", "")
        if s_env:
            try:
                s = float(s_env)
            except Exception:
                s = None
        else:
            # Heuristic smoothing ~ n * (noise_floor * std(y))^2
            noise_floor = float(os.environ.get("RKHS_NOISE_FLOOR", "0.01"))
            s = n * (noise_floor * float(np.std(y) + 1e-12)) ** 2
        try:
            spl = UnivariateSpline(t, y, k=5, s=s)
        except Exception:
            spl = UnivariateSpline(t, y, k=3, s=s)

        predictions = {}
        failures = {}
        for order in self.orders:
            try:
                d = spl.derivative(n=order)
                vals = d(self.x_eval)
                predictions[order] = [float(v) for v in np.asarray(vals)]
            except Exception as e:
                failures[order] = str(e)
                predictions[order] = [np.nan] * len(self.x_eval)
        return {"predictions": predictions, "failures": failures, "meta": {"s": s}}

    def _butterworth_spline(self) -> Dict:
        """Butterworth lowpass smoothing + quintic spline derivatives.

        Applies a 4th order Butterworth lowpass filter at 0.1 Nyquist frequency
        to smooth the signal, then computes derivatives via quintic spline.

        Returns:
            Dictionary containing predictions and failures from quintic spline
        """
        t = self.x_train
        y = self.y_train
        n = len(t)
        if n < 5:
            raise ValueError("Not enough points for Butterworth smoothing")
        # Design a 4th order low-pass at 0.1 Nyquist
        b, a = butter(N=4, Wn=0.1, btype='lowpass', analog=False)
        y_smooth = filtfilt(b, a, y)
        return self._quintic_spline_derivatives(t, y_smooth)

    def _finite_diff_spline(self) -> Dict:
        """Legacy FiniteDiff baseline replaced by smoothing + quintic-spline derivatives.

        Uses a milder lowpass filter (2nd order Butterworth at 0.2 Nyquist)
        to mimic a pre-filter prior to differentiation.

        Returns:
            Dictionary containing predictions and failures from quintic spline
        """
        # Use a milder lowpass to mimic a pre-filter prior to differentiation
        t = self.x_train
        y = self.y_train
        b, a = butter(N=2, Wn=0.2, btype='lowpass', analog=False)
        y_smooth = filtfilt(b, a, y)
        return self._quintic_spline_derivatives(t, y_smooth)

    def _svr_spline(self) -> Dict:
        """SVR (RBF) fit, then quintic-spline derivatives of predicted signal.

        Fits a Support Vector Regression model with RBF kernel to the training data,
        evaluates at x_eval, then computes derivatives via quintic spline.

        Returns:
            Dictionary containing predictions and failures from quintic spline
        """
        X = self.x_train.reshape(-1, 1)
        y = self.y_train
        svr = SVR(kernel='rbf', C=10.0, gamma='scale', epsilon=0.01)
        svr.fit(X, y)
        yhat_eval = svr.predict(self.x_eval.reshape(-1, 1))
        return self._quintic_spline_derivatives(self.x_eval, yhat_eval)
